File: 2021/day_4/main.go
package main

import (
	"adventofcode/adventUtils"
	"fmt"
	"strings"
)

type bingoPattern struct {
	numbers [5][5]int
	state   [5][5]bool
}

type bingoResult struct {
	p      bingoPattern
	rounds []int
}

type bit struct {
	zero int
	one  int
}

func removeEmpty(i []string) []string {
	r := []string{}
	for _, s := range i {
		if s != "" {
			r = append(r, s)
		}
	}
	return r
}

func parseLines(lines []string) ([]int, []bingoPattern) {
	firstLine := strings.Split(lines[0], ",")
	calledNumbers, err := adventUtils.ParseToInt(firstLine)
	if err != nil {
		// for _, e := range err {
		// 	fmt.Errorf("Error at %d: %+v\n", e.err, e.err)
		// }
		panic(err)
	}

	index := 2
	patterns := []bingoPattern{}
	for {
		if index > len(lines) {
			break
		}
		p := bingoPattern{}
		for j := 0; j < 5; j++ {
			nextLine := removeEmpty(strings.Split(lines[index+j], " "))
			bingoNumbers, err := adventUtils.ParseToInt(nextLine)
			if err != nil {
				panic(err)
			}
			copy(p.numbers[j][:], bingoNumbers[:5])
		}
		patterns = append(patterns, p)
		index += 6
	}

	return calledNumbers, patterns
}

func playBingo(called []int, patterns []bingoPattern) (bingoPattern, []int) {
	for i := 1; i <= len(called); i++ {
		s := called[:i]
		for _, p := range patterns {
			success, state := resolveBingo(s, p)
			p.state = state
			if success {
				return p, called[:i]
			}
		}
	}
	panic("Could not resolve Bingo")
}

func resolveBingo(n []int, p bingoPattern) (bool, [5][5]bool) {
	var state [5][5]bool
	for i := 0; i < len(p.numbers); i++ {
		for j := 0; j < len(p.numbers[i]); j++ {
			for _, c := range n {
				if p.numbers[i][j] == c {
					state[i][j] = true
					break
				}
			}
		}
	}
	for i := 0; i < len(p.numbers); i++ {
		if allTrue(state[i]) {
			return true, state
		}
		vert := [5]bool{state[0][i], state[1][i], state[2][i], state[3][i], state[4][i]}
		if allTrue(vert) {
			return true, state
		}
	}
	return false, state
}

func allTrue(i [5]bool) bool {
	for _, v := range i {
		if !v {
			return false
		}
	}
	return true
}

func calculateResult(p bingoPattern, r []int) int {
	left := 0
	for i := 0; i < len(p.numbers); i++ {
		for j := 0; j < len(p.numbers[i]); j++ {
			if !p.state[i][j] {
				left += p.numbers[i][j]
			}
		}
	}
	right := r[len(r)-1]

	return left * right
}

func loseBingo(called []int, patterns []bingoPattern) bingoResult {
	winners := []bingoResult{}
	winnerIndex := []int{}
	for i := 1; i <= len(called); i++ {
		s := called[:i]
		for j, p := range patterns {
			if adventUtils.IsIn(j, winnerIndex) {
				continue
			}
			success, state := resolveBingo(s, p)
			p.state = state
			if success {
				winners = append(winners, bingoResult{p: p, rounds: called[:i]})
				winnerIndex = append(winnerIndex, j)
			}
		}
	}

	return winners[len(winners)-1]
}

func day_4() {
	filename := "input.txt"
	lines, err := adventUtils.ReadFileAdvent(filename, 4)
	if err != nil {
		fmt.Errorf("%+v\n", err)
		panic(err)
	}
	calledNumbers, bingo := parseLines(lines)

	pattern, rounds := playBingo(calledNumbers, bingo)
	result := calculateResult(pattern, rounds) // Part 1
	fmt.Printf("Bingo was resolved after %d rounds. Result is %d\n", len(rounds), result)
	lastResult := loseBingo(calledNumbers, bingo)
	lostResult := calculateResult(lastResult.p, lastResult.rounds) // Part 1
	fmt.Printf("Bingo was list after %d rounds. Result is %d\n", len(lastResult.rounds), lostResult)
}

func main() {
	adventUtils.Benchmark(day_4)
}

File: 2021/day_4/main.rs
use std::collections::HashSet;
use std::convert::TryInto;
use std::fs;

#[derive(Debug, Copy, Clone)]
struct BingoPattern {
    numbers: [[i8; 5]; 5],
    state: [[bool; 5]; 5],
}

#[derive(Debug)]
struct BingoResult {
    pattern: BingoPattern,
    rounds: Vec<i8>,
}

fn read_file(path: &str) -> Vec<String> {
    let parsed = fs::read_to_string(path).expect("Could not read file");
    let split_lines = parsed.lines();
    split_lines.map(|s| s.to_string()).collect()
}

fn parse_lines(lines: &Vec<String>) -> (Vec<i8>, Vec<BingoPattern>) {
    let numbers_lines: Vec<i8> = lines[0].split(",").map(|s| s.parse().unwrap()).collect();

    let mut i = 2;
    let mut patterns = Vec::<BingoPattern>::new();
    while i < lines.len() {
        let mut numbers: [[i8; 5]; 5] = [[0; 5]; 5];
        for j in 0..5 {
            let line: Vec<i8> = lines[i + j]
                .split_ascii_whitespace()
                .map(|s| s.parse().unwrap())
                .collect();
            numbers[j] = line.try_into().expect("Slice with incorrect length");
        }
        let state = [[false; 5]; 5];
        patterns.push(BingoPattern {
            numbers: numbers,
            state: state,
        });
        i += 6;
    }
    return (numbers_lines, patterns);
}

fn resolve_bingo(numbers: &Vec<i8>, bingo: &BingoPattern) -> (bool, [[bool; 5]; 5]) {
    let mut state = [[false; 5]; 5];
    for i in 0..bingo.numbers.len() {
        for j in 0..bingo.numbers[0].len() {
            for value in numbers {
                if bingo.numbers[i][j] == *value {
                    state[i][j] = true;
                    break;
                }
            }
        }
    }

    for i in 0..bingo.numbers.len() {
        for _ in 0..bingo.numbers[0].len() {
            if state[i].iter().all(|x| *x == true) {
                return (true, state);
            }
            let vert = [
                state[0][i],
                state[1][i],
                state[2][i],
                state[3][i],
                state[4][i],
            ];
            if vert.iter().all(|x| *x == true) {
                return (true, state);
            }
        }
    }
    return (false, state);
}

fn calculate_result(pattern: &BingoPattern, numbers: &Vec<i8>) -> i32 {
    let mut left = 0i32;
    for i in 0..pattern.numbers.len() {
        for j in 0..pattern.numbers[0].len() {
            if !pattern.state[i][j] {
                left += i32::from(pattern.numbers[i][j]);
            }
        }
    }
    let right = i32::from(numbers[numbers.len() - 1]);

    return left * right;
}

fn play_normal_bingo(numbers: &Vec<i8>, patterns: &Vec<BingoPattern>) -> (BingoPattern, Vec<i8>) {
    for i in 5..numbers.len() {
        let number_slice = &numbers[0..i];
        for p in patterns {
            let (success, state) = resolve_bingo(&number_slice.to_vec(), p);
            if success {
                let result = BingoPattern {
                    numbers: p.numbers,
                    state: state,
                };
                return (result, number_slice.to_vec());
            }
        }
    }
    panic!("AAAAAH")
}

fn lose_bingo<'a>(numbers: &Vec<i8>, patterns: &Vec<BingoPattern>) -> BingoResult {
    let mut winner_index = HashSet::new();
    let mut last_winner: Option<BingoResult> = None;
    for i in 5..numbers.len() {
        let number_slice = &numbers[0..i];
        for (j, p) in patterns.iter().enumerate() {
            if winner_index.contains(&j) {
                continue;
            }
            let (success, state) = resolve_bingo(&number_slice.to_vec(), p);
            if success {
                last_winner = Some(BingoResult {
                    pattern: BingoPattern {
                        numbers: p.numbers,
                        state: state,
                    },
                    rounds: number_slice.to_vec(),
                });
                winner_index.insert(j);
            }
        }
    }
    last_winner.expect("Could not find any winners")
}

fn day_4() {
    let filename = "2021/day_4/input.txt";
    let lines = read_file(filename);
    let (numbers, patterns) = parse_lines(&lines);

    let (pattern, rounds) = play_normal_bingo(&numbers, &patterns);
    let result = calculate_result(&pattern, &rounds);
    println!(
        "Normal bingo was solved after {} rounds. The result is {}",
        rounds.len(),
        result
    );

    let last_result = lose_bingo(&numbers, &patterns);
    let lost_result = calculate_result(&last_result.pattern, &last_result.rounds);
    println!(
        "Losing bingo was solved after {} rounds. The result is {}",
        last_result.rounds.len(),
        lost_result
    );
}

fn main() {
    day_4();
}
